torch_xla/csrc/ops/expand_dynamic.cpp

```diff
@@ -0,0 +1,81 @@
+#include "torch_xla/csrc/ops/expand_dynamic.h"
+
+#include "absl/strings/str_join.h"
+#include "tensorflow/compiler/xla/client/lib/constants.h"
+#include "tensorflow/compiler/xla/xla_client/debug_macros.h"
+#include "torch/csrc/lazy/core/helpers.h"
+#include "torch/csrc/lazy/core/util.h"
+#include "torch_xla/csrc/data_ops.h"
+#include "torch_xla/csrc/lowering_context.h"
+#include "torch_xla/csrc/ops/infer_output_shape.h"
+
+namespace torch_xla {
+namespace {
+
+std::vector<xla::Shape> GetShapes(const torch::lazy::Value& input,
+                                  const std::vector<int64_t> upper_bounds,
+                                  const std::vector<bool> dynamic_dims) {
+  std::vector<xla::Shape> shapes;
+  shapes.push_back(GetXlaShape(input));
+  for (int i = 0; i < upper_bounds.size(); i++) {
+    shapes.push_back(
+        xla::ShapeUtil::MakeShape(GetXlaShape(input).element_type(),
+                                  {upper_bounds[i]}, {dynamic_dims[i]}));
+  }
+  return shapes;
+}
+
+xla::Shape NodeOutputShape(const torch::lazy::Value& input,
+                           const std::vector<int64_t> upper_bounds,
+                           const std::vector<bool> dynamic_dims) {
+  return xla::ShapeUtil::MakeShape(GetXlaShape(input).element_type(),
+                                   {upper_bounds}, {dynamic_dims});
+}
+
+std::vector<torch::lazy::Value> GetValues(
+    const torch::lazy::Value& input,
+    const std::vector<torch::lazy::Value> dimensions) {
+  std::vector<torch::lazy::Value> values = dimensions;
+  values.insert(values.begin(), input);
+  return values;
+}
+
+}  // namespace
+
+ExpandDynamic::ExpandDynamic(const torch::lazy::Value& input,
+                             const std::vector<torch::lazy::Value>& dimensions,
+                             const std::vector<int64_t> upper_bounds,
+                             const std::vector<bool> dynamic_dims,
+                             const torch::lazy::Shape& dynamic_shapes)
+    : XlaNode(
+          torch::lazy::OpKind(at::aten::expand), GetValues(input, dimensions),
+          {dynamic_shapes},
+          [&]() { return NodeOutputShape(input, upper_bounds, dynamic_dims); },
+          /*num_outputs=*/1, torch::lazy::MHash(upper_bounds)),
+      dynamic_shapes_(dynamic_shapes),
+      upper_bounds_(std::move(upper_bounds)),
+      dynamic_dims_(std::move(dynamic_dims)) {
+  shapes_ = GetShapes(input, upper_bounds_, dynamic_dims_);
+}
+
+XlaOpVector ExpandDynamic::Lower(LoweringContext* loctx) const {
+  xla::XlaOp input = loctx->GetOutputOp(operand(0));
+  std::vector<xla::XlaOp> size_ops;
+  for (int i = 1; i < shapes_.size(); i++) {
+    xla::XlaOp size_op = xla::Zeros(loctx->builder(), shapes_[i]);
+    size_ops.push_back(size_op);
+  }
+  xla::XlaOp output =
+      SetDimensionSizes(BuildExpand(input, upper_bounds_), size_ops);
```

Comment:
I think this part of logic is incorrect. You are actually set the dynamic size of input to a bunch of `0`. I think you mix the 
1. real size value (result of `getDimensionSize` for example, always a scalar)
2. whether dimension is dynamic

instead you pass a bunch of different shapes of `0` to the `SetDimensionSizes`.


Think of
```
t1 = tensor[10, 10, <=10]
size0 = getDimensionSize(t1, 0)
size1 = getDimensionSize(t1, 1)
size2 = getDimensionSize(t1, 2)

t2 = [1]
t2.expand(t1.static_size)
t2.setDimensionSize(size0, 0)         --> unnecessary
t2.setDimensionSize(size1, 1)          --> unnecessary
t2.setDimensionSize(size2, 2)
```

If you want to avoid unnecessary `setDimensionSize`, check whether `size0`, `size1` or `size2` are real size Node or just s int wrapped by symint (In the real code you always wrap `size` in a symint.)

Reply:
I addressed this note.